### ISSUE TITLE
Fails to build against Boost 1.66

### ISSUE THREAD

**Issue by jbeich** [arvidn/libtorrent#2585]
libtorrent version (or branch): master
platform/architecture: FreeBSD 11.1 amd64
compiler and compiler version: Clang/libc++ 4.0

I'm [preparing](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=223922) for Boost 1.66 which is planned to be released on 2017-12-09. boostorg/asio@b60e92b13ef6 broke libtorrent ([via](http://package22.nyi.freebsd.org/data/103amd64-default-PR223922/2017-12-02_11h06m39s/logs/errors/libtorrent-rasterbar-1.1.4_3.log)).

```c++
$ pkg install automake boost-libs git gmake libtool
$ git clone https://github.com/arvidn/libtorrent/
$ cd libtorrent
$ git describe --tags
libtorrent-1_1_5-1378-gc16eb32b9
$ ./autotool.sh
$ ./configure
$ gmake
[...]
In file included from bt_peer_connection.cpp:68:
In file included from ../include/libtorrent/socket_type.hpp:47:
In file included from ../include/libtorrent/ssl_stream.hpp:46:
In file included from /usr/local/include/boost/asio/ssl.hpp:22:
/usr/local/include/boost/asio/ssl/stream.hpp:85:39: error: no type named 'executor_type' in
      'libtorrent::utp_stream'
  typedef typename lowest_layer_type::executor_type executor_type;
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
../include/libtorrent/ssl_stream.hpp:78:19: note: in instantiation of template class
      'boost::asio::ssl::stream<libtorrent::utp_stream>' requested here
        typedef typename sock_type::next_layer_type next_layer_type;
                         ^
/usr/include/c++/v1/type_traits:1714:56: note: in instantiation of template class
      'libtorrent::ssl_stream<libtorrent::utp_stream>' requested here
                                                       __alignof__(_Types)...>::value;
                                                       ^
../include/libtorrent/socket_type.hpp:301:8: note: in instantiation of template class
      'std::__1::aligned_union<1, boost::asio::basic_stream_socket<boost::asio::ip::tcp>,
      libtorrent::socks5_stream, libtorrent::http_stream, libtorrent::utp_stream,
      libtorrent::i2p_stream,
      libtorrent::ssl_stream<boost::asio::basic_stream_socket<boost::asio::ip::tcp> >,
      libtorrent::ssl_stream<libtorrent::socks5_stream>,
      libtorrent::ssl_stream<libtorrent::http_stream>, libtorrent::ssl_stream<libtorrent::utp_stream>
      >' requested here
                aux::aligned_union<1
                     ^
1 error generated.
```

**Comment by arvidn** [comment]
does [this patch](https://github.com/arvidn/libtorrent/commit/f42b63c7ea82173837920d8712d2891cc3ecb6c2) fix it?
It kind of looks like a different issue though

**Comment by jbeich** [comment]
Still fails:
```c++
In file included from bt_peer_connection.cpp:68:
In file included from ../include/libtorrent/socket_type.hpp:47:
In file included from ../include/libtorrent/ssl_stream.hpp:46:
In file included from /usr/local/include/boost/asio/ssl.hpp:22:
/usr/local/include/boost/asio/ssl/stream.hpp:85:39: error: no type named 'executor_type' in
      'libtorrent::utp_stream'
  typedef typename lowest_layer_type::executor_type executor_type;
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~
../include/libtorrent/ssl_stream.hpp:78:19: note: in instantiation of template class
      'boost::asio::ssl::stream<libtorrent::utp_stream>' requested here
        typedef typename sock_type::next_layer_type next_layer_type;
                         ^
/usr/include/c++/v1/type_traits:1714:56: note: in instantiation of template class
      'libtorrent::ssl_stream<libtorrent::utp_stream>' requested here
                                                       __alignof__(_Types)...>::value;
                                                       ^
../include/libtorrent/socket_type.hpp:301:8: note: in instantiation of template class
      'std::__1::aligned_union<1, boost::asio::basic_stream_socket<boost::asio::ip::tcp>,
      libtorrent::socks5_stream, libtorrent::http_stream, libtorrent::utp_stream,
      libtorrent::i2p_stream,
      libtorrent::ssl_stream<boost::asio::basic_stream_socket<boost::asio::ip::tcp> >,
      libtorrent::ssl_stream<libtorrent::socks5_stream>,
      libtorrent::ssl_stream<libtorrent::http_stream>, libtorrent::ssl_stream<libtorrent::utp_stream>
      >' requested here
                aux::aligned_union<1
                     ^
1 error generated.
```

**Comment by jbeich** [comment]
@arvidn wrote:
> does this patch fix it?

f42b63c7ea82 is already included in c16eb32b94bb snapshot.

**Comment by jbeich** [comment]
Boost 1.66.0 rc1 is still affected, so the issue is unlikely to disappear on its own.

**Comment by jbeich** [comment]
Switching to GCC/libstdc++ 6.4.0 doesn't help:

```c++
In file included from /usr/local/include/boost/asio/ssl.hpp:22:0,
                 from ../include/libtorrent/ssl_stream.hpp:46,
                 from ../include/libtorrent/socket_type.hpp:47,
                 from bt_peer_connection.cpp:68:
/usr/local/include/boost/asio/ssl/stream.hpp: In instantiation of 'class boost::asio::ssl::stream<libtorrent::utp_stream>':
../include/libtorrent/ssl_stream.hpp:78:46:   required from 'class libtorrent::ssl_stream<libtorrent::utp_stream>'
/usr/local/lib/gcc6/include/c++/type_traits:2075:21:   recursively required from 'const size_t std::__strictest_alignment<libtorrent::socks5_stream, libtorrent::http_stream, libtorrent::utp_stream, libtorrent::i2p_stream, libtorrent::ssl_stream<boost::asio::basic_stream_socket<boost::asio::ip::tcp> >, libtorrent::ssl_stream<libtorrent::socks5_stream>, libtorrent::ssl_stream<libtorrent::http_stream>, libtorrent::ssl_stream<libtorrent::utp_stream> >::_S_size'
/usr/local/lib/gcc6/include/c++/type_traits:2075:21:   required from 'const size_t std::__strictest_alignment<boost::asio::basic_stream_socket<boost::asio::ip::tcp>, libtorrent::socks5_stream, libtorrent::http_stream, libtorrent::utp_stream, libtorrent::i2p_stream, libtorrent::ssl_stream<boost::asio::basic_stream_socket<boost::asio::ip::tcp> >, libtorrent::ssl_stream<libtorrent::socks5_stream>, libtorrent::ssl_stream<libtorrent::http_stream>, libtorrent::ssl_stream<libtorrent::utp_stream> >::_S_size'
/usr/local/lib/gcc6/include/c++/type_traits:2096:41:   required from 'const size_t std::aligned_union<1u, boost::asio::basic_stream_socket<boost::asio::ip::tcp>, libtorrent::socks5_stream, libtorrent::http_stream, libtorrent::utp_stream, libtorrent::i2p_stream, libtorrent::ssl_stream<boost::asio::basic_stream_socket<boost::asio::ip::tcp> >, libtorrent::ssl_stream<libtorrent::socks5_stream>, libtorrent::ssl_stream<libtorrent::http_stream>, libtorrent::ssl_stream<libtorrent::utp_stream> >::_S_len'
/usr/local/lib/gcc6/include/c++/type_traits:2102:71:   required from 'struct std::aligned_union<1u, boost::asio::basic_stream_socket<boost::asio::ip::tcp>, libtorrent::socks5_stream, libtorrent::http_stream, libtorrent::utp_stream, libtorrent::i2p_stream, libtorrent::ssl_stream<boost::asio::basic_stream_socket<boost::asio::ip::tcp> >, libtorrent::ssl_stream<libtorrent::socks5_stream>, libtorrent::ssl_stream<libtorrent::http_stream>, libtorrent::ssl_stream<libtorrent::utp_stream> >'
../include/libtorrent/socket_type.hpp:315:4:   required from here
/usr/local/include/boost/asio/ssl/stream.hpp:85:53: error: no type named 'executor_type' in 'boost::asio::ssl::stream<libtorrent::utp_stream>::lowest_layer_type {aka struct libtorrent::utp_stream}'
   typedef typename lowest_layer_type::executor_type executor_type;
                                                     ^~~~~~~~~~~~~
```

**Comment by arvidn** [comment]
Does this patch fix it? https://github.com/arvidn/libtorrent/pull/2601

**Comment by jbeich** [comment]
No, see [build log](http://sprunge.us/AfZV).

**Comment by arvidn** [comment]
Unfortunately, at least one of these errors appear to be a regression in asio. I filed an issue here: https://github.com/boostorg/asio/issues/74

**Comment by arvidn** [comment]
never mind, that was an intentional change in asio. I've updated the PR. it builds with openssl enabled for me now.

**Comment by jbeich** [comment]
I confirm, [builds fine](http://sprunge.us/cFZi) after #2601.

**Comment by arvidn** [comment]
this patch has been merged into RC_1_1 now
